src/datachain/asyn.py

```diff
@@ -179,6 +180,7 @@ def iterate(self, timeout=None) -> Generator[ResultT, None, None]:
             self.shutdown_producer()
             if not async_run.done():
                 async_run.cancel()
+                wait([async_run])
```

Comment:
`.cancel()` does not immediately cancel the task the underlying `asyncio` task.

We could add a `.result()` to wait for the future, but that does not seem to work either for the cancelled future from `run_coroutine_threadsafe()`. See https://github.com/python/cpython/issues/105836.

So, I have added `wait(...)` as it seems to wait the cancelled future, and wait for underlying asyncio task.

Alternatively, we could add an `asyncio.Event` and wait for it.